core-bundle/src/Resources/contao/library/Contao/StringUtil.php

```diff
@@ -971,6 +971,78 @@ public static function ampersand($strString, $blnEncode=true): string
 	{
 		return preg_replace('/&(amp;)?/i', ($blnEncode ? '&amp;' : '&'), $strString);
 	}
+
+	/**
+	 * Converts an input-encoded string back to its raw UTF-8 value it originated from.
+	 *
+	 * It handles all Contao input encoding specifics like insert tags, basic entities and encoded entities.
+	 *
+	 * @param bool $blnRemoveInsertTags True to remove insert tags instead of replacing them
+	 */
+	public static function getRawDecodedValue(string $strValue, bool $blnRemoveInsertTags = false): string
+	{
+		if ($blnRemoveInsertTags)
+		{
+			$strValue = static::stripInsertTags($strValue);
+		}
+		else
+		{
+			$strValue = Controller::replaceInsertTags($strValue, false);
+		}
+
+		$strValue = strip_tags($strValue);
+		$strValue = static::restoreBasicEntities($strValue);
+		$strValue = static::decodeEntities($strValue);
+
+		// Ensure valid UTF-8
+		if (preg_match('//u', $strValue) !== 1)
+		{
+			$substituteCharacter = mb_substitute_character();
+			mb_substitute_character(0xFFFD);
+
+			$strValue = mb_convert_encoding($strValue, 'UTF-8', 'UTF-8');
+
+			mb_substitute_character($substituteCharacter);
+		}
+
+		// TODO: We have to prevent insert tags from existing in this stage for security reasons, is using a ZWNJ the best we can do?
+		$strValue = str_replace('{', "\u{200C}{", $strValue);
```

Comment:
How can they still exist here? They are either stripped or replaced, no?

Reply:
Encoded curly braces would end up here. There is a unit test for this case.

Reply:
Ah, yeah I mean they were either stripped or replaced before which means after we decoded, we can strip again here because that's not supported, is it? Afaik we have this `[{]` notation if you want to output curly braces whithout having them interpreted?